File: app/src/main/java/com/close/hook/ads/hook/preference/PreferencesHelper.java
package com.close.hook.ads.hook.preference;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.Optional;

import de.robv.android.xposed.XSharedPreferences;

public class PreferencesHelper {

    private static final String PREF_NAME = "com.close.hook.ads_preferences";

    private final Optional<SharedPreferences> prefs;
    private final Optional<XSharedPreferences> xPrefs;

/*
    public PreferencesHelper(Context context) {
        prefs = Optional.of(context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE));
        xPrefs = Optional.empty();

        try {
            File prefFile = new File("/data/data/" + context.getPackageName() + "/shared_prefs/" + PREF_NAME + ".xml");
            if (prefFile.exists()) {
                prefFile.setReadable(true, false);
            }
        } catch (Exception ignored) {
        }
    }
*/

    public PreferencesHelper(Context context) {
        prefs = Optional.of(context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE));
        xPrefs = Optional.empty();
    }

    public PreferencesHelper() {
        xPrefs = Optional.of(new XSharedPreferences("com.close.hook.ads", PREF_NAME));
        prefs = Optional.empty();
        xPrefs.ifPresent(xp -> {
            xp.makeWorldReadable();
            xp.reload();
        });
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        if (prefs.isPresent()) {
            return prefs.get().getBoolean(key, defaultValue);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().getBoolean(key, defaultValue);
        }
        return defaultValue;
    }

    public void setBoolean(String key, boolean value) {
        prefs.ifPresent(p -> p.edit().putBoolean(key, value).apply());
    }

    public int getInt(String key, int defaultValue) {
        if (prefs.isPresent()) {
            return prefs.get().getInt(key, defaultValue);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().getInt(key, defaultValue);
        }
        return defaultValue;
    }

    public void setInt(String key, int value) {
        prefs.ifPresent(p -> p.edit().putInt(key, value).apply());
    }

    public long getLong(String key, long defaultValue) {
        if (prefs.isPresent()) {
            return prefs.get().getLong(key, defaultValue);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().getLong(key, defaultValue);
        }
        return defaultValue;
    }

    public void putLong(String key, long value) {
        prefs.ifPresent(p -> p.edit().putLong(key, value).apply());
    }

    public float getFloat(String key, float defaultValue) {
        if (prefs.isPresent()) {
            return prefs.get().getFloat(key, defaultValue);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().getFloat(key, defaultValue);
        }
        return defaultValue;
    }

    public void setFloat(String key, float value) {
        prefs.ifPresent(p -> p.edit().putFloat(key, value).apply());
    }

    public String getString(String key, String defaultValue) {
        if (prefs.isPresent()) {
            return prefs.get().getString(key, defaultValue);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().getString(key, defaultValue);
        }
        return defaultValue;
    }

    public void setString(String key, String value) {
        prefs.ifPresent(p -> p.edit().putString(key, value).apply());
    }

    public boolean contains(String key) {
        if (prefs.isPresent()) {
            return prefs.get().contains(key);
        } else if (xPrefs.isPresent()) {
            xPrefs.get().reload();
            return xPrefs.get().contains(key);
        }
        return false;
    }

    public void remove(String key) {
        prefs.ifPresent(p -> p.edit().remove(key).apply());
    }

    public void clear() {
        prefs.ifPresent(p -> p.edit().clear().apply());
    }
}

File: app/src/main/java/com/close/hook/ads/hook/preference/PreferencesHelper.kt
package com.close.hook.ads.hook.preference

import android.content.Context
import android.content.SharedPreferences
import de.robv.android.xposed.XSharedPreferences

class PreferencesHelper {

    companion object {
        private const val PREF_NAME = "com.close.hook.ads_preferences"
        private const val MODULE_PACKAGE_NAME = "com.close.hook.ads"
    }

    private val prefs: SharedPreferences?
    private val xPrefs: XSharedPreferences?

    constructor(context: Context) {
        this.prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        this.xPrefs = null
    }

    constructor() {
        this.prefs = null
        this.xPrefs = XSharedPreferences(MODULE_PACKAGE_NAME, PREF_NAME).also {
            it.makeWorldReadable()
            it.reload()
        }
    }

    fun reloadXPrefs() {
        xPrefs?.reload()
    }

    fun getBoolean(key: String, defaultValue: Boolean): Boolean {
        return prefs?.getBoolean(key, defaultValue)
            ?: xPrefs?.getBoolean(key, defaultValue)
            ?: defaultValue
    }

    fun setBoolean(key: String, value: Boolean) {
        prefs?.edit()?.putBoolean(key, value)?.apply()
    }

    fun getInt(key: String, defaultValue: Int): Int {
        return prefs?.getInt(key, defaultValue)
            ?: xPrefs?.getInt(key, defaultValue)
            ?: defaultValue
    }

    fun setInt(key: String, value: Int) {
        prefs?.edit()?.putInt(key, value)?.apply()
    }

    fun getLong(key: String, defaultValue: Long): Long {
        return prefs?.getLong(key, defaultValue)
            ?: xPrefs?.getLong(key, defaultValue)
            ?: defaultValue
    }

    fun setLong(key: String, value: Long) {
        prefs?.edit()?.putLong(key, value)?.apply()
    }

    fun getFloat(key: String, defaultValue: Float): Float {
        return prefs?.getFloat(key, defaultValue)
            ?: xPrefs?.getFloat(key, defaultValue)
            ?: defaultValue
    }

    fun setFloat(key: String, value: Float) {
        prefs?.edit()?.putFloat(key, value)?.apply()
    }

    fun getString(key: String, defaultValue: String): String {
        val result = prefs?.getString(key, null)
        return if (result != null) result
               else xPrefs?.getString(key, defaultValue) ?: defaultValue
    }

    fun setString(key: String, value: String) {
        prefs?.edit()?.putString(key, value)?.apply()
    }

    fun contains(key: String): Boolean {
        return prefs?.contains(key)
            ?: xPrefs?.contains(key)
            ?: false
    }

    fun remove(key: String) {
        prefs?.edit()?.remove(key)?.apply()
    }

    fun clear() {
        prefs?.edit()?.clear()?.apply()
    }
}